### PR TITLE
Transactions API data

### PR DESCRIPTION
**Pull request by samerbuna**
- Refactor WalletTransaction types
- Abstract common props into 'baseTransaction' in 'fromLedger'
- Make new 'ExtendedLedgerTransactionType' type
- Refactor 'fromLedger' to reflect new WalletTransaction types
- Refactor 'filterPendingIncoming to return new LedgerTransaction types
- Refactor new LedgerTransaction types into old graphql server
- Refactor initiationVia/settlementVia in graphql objects
- Refactor new LedgerTransaction types into tests
- Temporary: Set PaymentSecret to null
- Implement GraphQL types for the new InitiationVia/SettlementVia unions

### CODE REVIEW

File: src/domain/wallets/index.types.d.ts [hunk 1]
```diff
@@ -12,11 +10,47 @@ type Deprecated = {
...
+  readonly paymentHash: PaymentHash
+  readonly pubkey: Pubkey
+}
+
+type InitiationViaOnChain = {
+  readonly type: PaymentInitiationMethod["OnChain"]
+  readonly address: OnChainAddress
+}
+
+type InitiationViaOnChainLegacy = {
```
Comment (nicolasburtey) [inline src/domain/wallets/index.types.d.ts:30]: why is that needed?

File: src/domain/wallets/index.types.d.ts [hunk 2]
```diff
@@ -12,11 +10,47 @@ type Deprecated = {
...
+
+type SettlementViaIntraledger = {
+  readonly type: SettlementMethod["IntraLedger"]
+  readonly walletId: WalletId
+  readonly counterPartyUsername: Username | null
+}
+
+type SettlementViaLn = {
+  readonly type: SettlementMethod["Lightning"]
+  paymentSecret: PaymentSecret | null
```
Comment (nicolasburtey) [inline src/domain/wallets/index.types.d.ts:43]: this shouldn't be `null`? we should always know the `paymentSecret`  once this has been settled? maybe the reason this may not be `null` is if the payment is in flight?

File: src/domain/wallets/tx-history.ts [hunk 3]
```diff
@@ -17,23 +17,26 @@ const filterPendingIncoming = (
...
-          otherPartyUsername: null,
-          settlementFee: toSats(0),
-          transactionHash: rawTx.txHash,
-          status: TxStatus.Pending,
-          memo: null,
-          createdAt: createdAt,
-          settlementAmount: sats,
-          settlementUsdPerSat: usdPerSat,
-          address,
+          initiationVia: {
```
Comment (nicolasburtey) [inline src/domain/wallets/tx-history.ts:32]: if this is `initiationVia` without an appendix telling which protocol we are using, I would assume this would cover all the initiation option, yet we have only have `OnChain` in the object below?

File: src/domain/wallets/tx-history.ts [hunk 4]
```diff
@@ -65,92 +68,143 @@ export const fromLedger = (
...
+          walletTransaction = {
+            ...baseTransaction,
+            initiationVia: {
+              type: PaymentInitiationMethod.Lightning,
+              paymentHash: paymentHash as PaymentHash,
+              pubkey: pubkey as Pubkey,
+            },
+            settlementVia: {
+              type: SettlementMethod.Lightning,
+              paymentSecret: null,
```
Comment (nicolasburtey) [inline src/domain/wallets/tx-history.ts:185]: should not be `null`?

Comment (nicolasburtey) [reply src/domain/wallets/tx-history.ts:185]: `paymentSecret` should probably be treated differently whether it's a `Payment` of an `Invoice`:

- in the case of an invoice, "we" (ie: our lnd node) actually know the `secret` right from the beginning because we created it. but we should not disclosed it before the invoice is being paid, because we don't want to let a end user know/share it until, this would be an attack vector. ie: an intermediary node could know the secret before it reaches our node, and the payment would never arrived to us so our user would not be credited, but the payer would have sent the money. maybe not a high risk scenario but something to consider.

- in the case of a payment, when we know the secret, it means the invoice has been settled, so there is no risk here

File: src/domain/wallets/tx-history.ts [hunk 5]
```diff
@@ -65,92 +68,143 @@ export const fromLedger = (
...
+            },
+          }
+          return walletTransaction
+      }
+
+      walletTransaction = {
```
Comment (nicolasburtey) [inline src/domain/wallets/tx-history.ts:191]: shouldn't this also be part of the `case:` scenario?

Comment (nicolasburtey) [reply src/domain/wallets/tx-history.ts:191]: that would be cleaner that assuming a default?

**Review by nicolasburtey — COMMENTED** [review]
couple of comments/questions

File: src/domain/wallets/index.types.d.ts (hunk 1 above)
Comment (vindard) [reply src/domain/wallets/index.types.d.ts:30]: This was for onchain transaction that have `undefined` address in the db for older version of the codebase that didn't persist this

File: src/domain/wallets/index.types.d.ts (hunk 2 above)
Comment (vindard) [reply src/domain/wallets/index.types.d.ts:43]: This was a temporary thing because we don't persist this yet for successful payments. I'm planning to add this in after the persist-payments PR (#665) is merged.

File: src/domain/wallets/tx-history.ts (hunk 4 above)
Comment (vindard) [reply src/domain/wallets/tx-history.ts:185]: True, I can make another type to reflect this (`WalletLnSettledSentTransaction` & `WalletLnSettledReceivedTransaction`) with different names for that property (`paymentSecretGenerated` vs. `paymentSecretReceived` for e.g.)

File: src/domain/wallets/tx-history.ts (hunk 3 above)
Comment (vindard) [reply src/domain/wallets/tx-history.ts:32]: We differentiate this via the `type` sub-property inside `initiationVia`, not the actual property name. In this case though, I could actually make this more accurate by changing the return type of the `filterPendingIncoming` to be `WalletOnChainTransaction`.

File: src/domain/wallets/tx-history.ts (hunk 5 above)
Comment (vindard) [reply src/domain/wallets/tx-history.ts:191]: This part of the code was a little messy and needs to be cleaned up a bit. It's tricky because we have send/receive types mixed in with things like `"routing_fee"` and `"exchange_rebalance"` under the same `LedgerTransactionType` definition. I tried a few things but ended up in some dead-ends, so opted to maintain the current behaviour which defaults to Intraledger initiated and settled and then to clean this up in another later PR (to not block the rest of Samer's integration work).

File: src/domain/wallets/tx-history.ts (hunk 4 above)
Comment (vindard) [reply src/domain/wallets/tx-history.ts:185]: > should not be `null`?

See this [comment](https://github.com/GaloyMoney/galoy/pull/759#discussion_r757837347)

File: src/domain/wallets/index.types.d.ts [hunk 6]
```diff
@@ -12,11 +10,47 @@ type Deprecated = {
...
+  readonly type: PaymentInitiationMethod["Lightning"]
+  readonly paymentHash: PaymentHash
+  readonly pubkey: Pubkey
+}
+
+type InitiationViaOnChain = {
+  readonly type: PaymentInitiationMethod["OnChain"]
+  readonly address: OnChainAddress
+}
+
```
Comment (nicolasburtey) [inline src/domain/wallets/index.types.d.ts:29]:
```suggestion
FIXME: add OnChainAddress associated with old transaction to remove this legacy type
```

File: src/graphql/types/abstract/settlement-via.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+const SettlementViaIntraLedger = new GT.Object({
+  name: "SettlementViaIntraLedger",
+  fields: () => ({
+    walletId: {
+      type: GT.NonNull(WalletId),
+    },
+    counterPartyUsername: {
+      type: Username,
+      description: dedent`Settlement destination: Could be null the payee does not have a username`,
```
Comment (nicolasburtey) [inline src/graphql/types/abstract/settlement-via.ts:20]:
```suggestion
      description: dedent`Settlement destination: Could be null if the payee does not have a username`,
```

File: src/graphql/types/object/transaction.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+      type: GT.NonNull(InitiationVia),
+      description: "From which protocol the payment has been initiated.",
+    },
+    settlementVia: {
+      type: GT.NonNull(SettlementVia),
+      description: "To which protocol the payment has settled on.",
+    },
+    settlementAmount: {
+      type: GT.NonNull(SatAmount),
+      description: "Amount of sats paid by the user.",
```
Comment (nicolasburtey) [inline src/graphql/types/object/transaction.ts:36]: this is inclusive fees? maybe should be added to the description?

File: src/graphql/types/abstract/initiation-via.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+      type: Username,
+    },
+  }),
+})
+
+const InitiationViaLn = new GT.Object({
+  name: "InitiationViaIntraLn",
+  fields: () => ({
+    paymentRequest: {
+      type: GT.NonNull(LnPaymentRequest),
```
Comment (nicolasburtey) [inline src/graphql/types/abstract/initiation-via.ts]: sometime the amount is mandatory (amountless invoice), sometimes it's explicitly forbidden. how are we dealing with this here?

because amount is somehow related to the initiation type

File: src/domain/wallets/tx-history.ts (hunk 5 above)
Comment (nicolasburtey) [reply src/domain/wallets/tx-history.ts:191]: no strong opinion on this except that `"routing_fee"` and `"exchange_rebalance"` are not meant to be returned to any normal user.

it's a transaction that is related to the admin/"bank owner", so this won't be expose to the public API in my mind.

Comment (vindard) [reply src/domain/wallets/tx-history.ts:191]: Yup exactly, we should probably differentiate those into another domain type

File: src/domain/wallets/index.types.d.ts (hunk 6 above)
Comment (vindard) [reply src/domain/wallets/index.types.d.ts]: I added this commit, but thinking about it now I'm not sure if we'd have this information anywhere else to backfill for transactions that were initiated onchain but settled intraledger 🤔

File: src/domain/wallets/tx-history.ts [hunk 10]
```diff
@@ -65,92 +68,143 @@ export const fromLedger = (
...
+      }
+
+      let walletTransaction: WalletTransaction
+      switch (txType) {
+        case ExtendedLedgerTransactionType.IntraLedger:
+          walletTransaction = {
+            ...baseTransaction,
+            initiationVia: {
+              type: PaymentInitiationMethod.IntraLedger,
+              walletId: walletId as WalletId,
```
Comment (dolcalmi) [inline src/domain/wallets/tx-history.ts:117]: why do we need walletId here? is the same base walletId

File: src/graphql/types/object/transaction.ts (hunk 8 above)
Comment (dolcalmi) [reply src/graphql/types/object/transaction.ts:36]: yes, this amount includes fee

image

**Review by dolcalmi — APPROVED** [review]
as reference, this issues need to be solved after merge

- Add counterPartyWalletId (publicid) to medici transactions https://github.com/GaloyMoney/galoy/issues/765
- Add payment secret for invoices and payments https://github.com/GaloyMoney/galoy/issues/764

File: src/domain/ledger/index.ts [hunk 11]
```diff
@@ -25,6 +26,11 @@ export const LedgerTransactionType = {
   ToHotWallet: "to_hot_wallet",
 } as const
 
+export const ExtendedLedgerTransactionType = {
```
Comment (bodymindarts) [inline src/domain/ledger/index.ts:29]: Why was this added as an additional type.

**Review by bodymindarts — APPROVED** [review]

File: src/graphql/types/abstract/initiation-via.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+import OnChainAddress from "../scalar/on-chain-address"
+import { PaymentInitiationMethod } from "@domain/wallets"
+import PaymentHash from "../scalar/payment-hash"
+
+const InitiationViaIntraLedger = new GT.Object({
+  name: "InitiationViaIntraLedger",
+  isTypeOf: (source) => source.type === PaymentInitiationMethod.IntraLedger,
+  fields: () => ({
+    counterPartyWalletId: {
+      // type: GT.NonNull(WalletId),
```
Comment (vindard) [inline src/graphql/types/abstract/initiation-via.ts:14]: Based on https://github.com/GaloyMoney/galoy/pull/771#issuecomment-983855316 should I change this back as well?

File: src/domain/ledger/index.ts (hunk 11 above)
Comment (vindard) [reply src/domain/ledger/index.ts:29]: It was to help us distinguish ln-intraledger from normal intraledger transactions since we currently lump both under `on_us` in the db whereas onchain-intraledger has its own `onchain_on_us` type.

It got used in the `tx-history` switch when categorizing different transactions into their respective `initiationVia` and `settlementVia` types.
